src/troykaph.cpp

```diff
@@ -0,0 +1,89 @@
+/*
+ * This file is a part of TroykaPH library.
+ *
+ * Product page: https://amperka.ru/product/zelo-folow-line-sensor
+ * © Amperka LLC (https://amperka.com, dev@amperka.com)
+ * 
+ * Author: Yury Botov <by@amperka.com>
+ * License: GPLv3, all text here must be included in any redistribution.
+ */
+
+#include "TroykaPH.h"
+
+TroykaPH::TroykaPH(uint8_t pin) {
+    _pin = pin;
+    _lastValue = 7;
+    _correction = 1.0;
+    _nextMeasureTime = 0;
+}
+
+void TroykaPH::begin(float correction, float zeroLevel) {
+    _correction = correction;
+    _zeroLevel = zeroLevel;
+}
+
+static float fmap(float value, float minVal, float maxVal, float minOut, float maxOut) {
+    return (value - minVal) / (maxVal - minVal) * (maxOut - minOut) + minOut;
+}
+
+void TroykaPH::update(long periodMilliseconds) {
+    constexpr float idealVcc = 5.0;
+    constexpr float minPh = 0.0;
+    constexpr float maxPh = 14.0;
+    constexpr float phHalfRangeInVolts = 0.82;
+
+    float value = 0;
+    if (millis() - _nextMeasureTime > periodMilliseconds) {
+        _nextMeasureTime += periodMilliseconds;
+        // read value
+        (void)analogRead(_pin);
```

Comment:
Googling showed that in cases where greater accuracy is needed, you must first make an empty read so that the internal analog multiplexer switches to the desired channel, and only after a pause, when the internal capacitor voltage value stabilizes, read real data.